zen/client.go

```diff
@@ -549,29 +549,43 @@ func (b *Client) parseTransactions(
 		coins = addCoinsFromSameBlock(tx.Operations, coins)
 	}
 
-	for index, certificate := range block.MaturedCerts {
+	for index, matureCertificate := range block.MaturedCerts {
 		// For matured certificates, we only parse outputs that are backward transfers
 		backwardTransferOutputs := []*Output{}
-
-		for i := range certificate.Outputs {
-			if certificate.Outputs[i].BackwardTransfer == true  {
-				backwardTransferOutputs = append(backwardTransferOutputs, certificate.Outputs[i])
+		
+		for i := range matureCertificate.Outputs {
+			
+			if matureCertificate.Outputs[i].BackwardTransfer  {
+				backwardTransferOutputs = append(backwardTransferOutputs, matureCertificate.Outputs[i])
 			}
 		}
 
-		certTxOps, err := b.parseTxOperations([]*Input{}, backwardTransferOutputs, certificate.Hash, len(block.Txs) + len(block.Certs) + index, coins, false)
+		matureCertTxOps, err := b.parseTxOperations([]*Input{}, backwardTransferOutputs, matureCertificate.Hash, len(block.Txs) + len(block.Certs) + index, coins, false)
 		if err != nil {
 			return nil, fmt.Errorf("%w: error parsing mature certificate transaction operations", err)
 		}
 
+		certAndMatureCertTogether:=false 
+		for _, tx := range txs {
+			if tx.TransactionIdentifier.Hash == matureCertificate.Hash {
+				for _, matureCertTxOp := range matureCertTxOps {
+					matureCertTxOp.OperationIdentifier.Index += int64(len(tx.Operations)) 
```

Comment:
Could you explain what is being done here? Specifically why the number of operations is important

Reply:
Sure thing, @yurikoinaba. txs is a list of all the txs in the block including the ones that are in the cert and matureCert fields. The main problem we are resolving is that the same certificate tx can be in the cert and matureCert fields and right now rosetta-zen creates two separate entries in that txs list. But we just want one. The cert txs have already been added to the txs list so here we look through the matureCert txs and see if any of them are for the same cert. If one of them is we are only interested in the backward transfer outputs. Operations is rosetta-zen's name for both  the input and output boxes of a given tx, and since we are adding the backward transfer output boxes to the already existing cert tx we ensure it gets the right index number by adding the length of that existing cert tx's operations list. Without this logic, each of the backward transfer output boxes would be added to the operations list with erroneous, duplicated indexes of 0,1,2, etc.